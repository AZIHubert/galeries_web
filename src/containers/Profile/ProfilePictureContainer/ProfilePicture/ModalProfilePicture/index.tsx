import moment from 'moment';
import * as React from 'react';

import { UserContext } from '#contexts/UserContext';

interface ModalProfilPictureI {
  profilePicture: ProfilePictureI;
}

const formatBytes = (a: number, b = 2) => {
  if (a === 0) return '0 Bytes'; const c = b < 0 ? 0 : b;
  const d = Math.floor(Math.log(a) / Math.log(1024));
  return `${parseFloat((a / 1024 ** d).toFixed(c))} ${['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'][d]}`;
};

const ModalProfilPicture = ({
  profilePicture,
}: ModalProfilPictureI) => {
  const {
    createdAt,
    originalImage: {
      height,
      signedUrl,
      size,
      width,
    },
  } = profilePicture;
  const { setUser, user } = React.useContext(UserContext);
  const switchCurrent = () => {
    setUser((prevState) => {
      if (prevState) {
        const { currentProfilePicture } = prevState;
        if (currentProfilePicture && currentProfilePicture.id === profilePicture.id) {
          return {
            ...prevState,
            currentProfilePicture: null,
            currentProfilePictureId: null,
          };
        }
        return {
          ...prevState,
          currentProfilePicture: profilePicture,
          currentProfilePictureId: profilePicture.id,
        };
      }
      return null;
    });
  };
  const current = user
    ? user.currentProfilePictureId === profilePicture.id
    : null;
  return (
    <div>
      <img src={signedUrl} alt='image' />
      <p>
        upload at
      </p>
      <p
        data-testid='uploadAtText'
      >
        {` ${moment(createdAt).format('MMMM Do YYYY')}`}
      </p>
      <p>
        weight:
      </p>
      <p
        data-testid='weightText'
      >
        {formatBytes(size)}
      </p>
      <p>
        size:
      </p>
      <p
        data-testid='sizeText'
      >
        {`${width} x ${height} px`}
      </p>
      <button
        data-testid='modalProfilePictureButton'
        onClick={() => switchCurrent()}
      >
        {current ? 'remove profile picture' : 'use as profile picture'}
      </button>
      <button>
        download image
      </button>
    </div>
  );
};

export default ModalProfilPicture;
