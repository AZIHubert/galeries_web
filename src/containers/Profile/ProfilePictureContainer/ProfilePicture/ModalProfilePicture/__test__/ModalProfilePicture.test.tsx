import {
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import * as React from 'react';
import renderer from 'react-test-renderer';

import {
  ImageI,
  ProfilePictureI,
} from '#helpers/interfaces';

import ModalProfilePicture from '../index';

const newImage: ImageI = {
  bucketName: 'bucketName',
  fileName: 'fileName',
  format: 'jpg',
  height: 2340,
  id: '1',
  signedUrl: 'http://url.com',
  size: 100857,
  width: 1080,
};
const newProfilePicture: ProfilePictureI = {
  createdAt: new Date('2021-02-05 18:42:02.528+01'),
  cropedImage: newImage,
  id: '1',
  originalImage: newImage,
  pendingImage: newImage,
};

const Container = ({
  defaultCurrent = false,
}: {
  defaultCurrent?: boolean
}) => {
  const [current, setCurrent] = React.useState<boolean>(defaultCurrent);
  const switchCurrent = () => {
    setCurrent((prevState) => !prevState);
  };
  return (
    <ModalProfilePicture
      current={current}
      profilePicture={newProfilePicture}
      switchCurrent={switchCurrent}
    />
  );
};

describe('Body', () => {
  const mockSetCurrent = jest.fn();
  afterEach(cleanup);
  it('renders without crashing', () => {
    const tree = renderer.create(
      <ModalProfilePicture
        profilePicture={newProfilePicture}
        switchCurrent={mockSetCurrent}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should format date properly', () => {
    const { getByTestId } = render(<Container />);
    const uploadAtText = getByTestId('uploadAtText');
    expect(uploadAtText).toHaveTextContent('February 5th 2021');
  });
  it('should format weight properly', () => {
    const { getByTestId } = render(<Container />);
    const weightText = getByTestId('weightText');
    expect(weightText).toHaveTextContent('98.49 KB');
  });
  it('should format size properly', () => {
    const { getByTestId } = render(<Container />);
    const sizeText = getByTestId('sizeText');
    expect(sizeText).toHaveTextContent('1080 x 2340 px');
  });
  it('should display profilePicture button properly if not current', () => {
    const { getByTestId } = render(<Container />);
    const profilePictureButton = getByTestId('profilePictureButton');
    expect(profilePictureButton).toHaveTextContent('use as profile picture');
  });
  it('should display profilePicture button properly if current', () => {
    const { getByTestId } = render(<Container
      defaultCurrent={true}
    />);
    const profilePictureButton = getByTestId('profilePictureButton');
    expect(profilePictureButton).toHaveTextContent('remove profile picture');
  });
  it('should switch current when clicking on profilePicture button', () => {
    const { getByTestId } = render(<Container />);
    const profilePictureButton = getByTestId('profilePictureButton');
    fireEvent.click(profilePictureButton);
    const profilePictureButtonWithCurrentTrue = screen
      .getByTestId('profilePictureButton');
    expect(profilePictureButtonWithCurrentTrue)
      .toHaveTextContent('remove profile picture');
    fireEvent.click(profilePictureButton);
    const profilePictureButtonWithCurrentFalse = screen
      .getByTestId('profilePictureButton');
    expect(profilePictureButtonWithCurrentFalse)
      .toHaveTextContent('use as profile picture');
  });
});
