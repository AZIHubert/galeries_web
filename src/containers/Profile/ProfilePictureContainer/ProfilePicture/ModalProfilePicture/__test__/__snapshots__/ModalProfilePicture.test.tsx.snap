// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body renders without crashing 1`] = `
<div>
  <img
    alt="image"
    src="http://url.com"
  />
  <p>
    upload at
  </p>
  <p
    data-testid="uploadAtText"
  >
     February 5th 2021
  </p>
  <p>
    weight:
  </p>
  <p
    data-testid="weightText"
  >
    98.49 KB
  </p>
  <p>
    size:
  </p>
  <p
    data-testid="sizeText"
  >
    1080 x 2340 px
  </p>
  <button
    data-testid="profilePictureButton"
    onClick={[MockFunction]}
  >
    use as profile picture
  </button>
  <button>
    download image
  </button>
</div>
`;
